cdc/generation.cc

```diff
@@ -928,17 +928,17 @@ future<bool> generation_service::do_handle_cdc_generation(db_clock::time_point t
             ts, db_clock::now()));
     }
 
-    // If we're not gossiping our own generation timestamp (because we've upgraded from a non-CDC/old version,
-    // or we somehow lost it due to a byzantine failure), start gossiping someone else's timestamp.
-    // This is to avoid the upgrade check on every restart (see `should_propose_first_cdc_generation`).
-    // And if we notice that `ts` is higher than our timestamp, we will start gossiping it instead,
-    // so if the node that initially gossiped `ts` leaves the cluster while `ts` is still the latest generation,
-    // the cluster will remember.
-    if (!_gen_ts || *_gen_ts < ts) {
-        _gen_ts = ts;
-        co_await db::system_keyspace::update_cdc_streams_timestamp(ts);
+    // We always gossip about the generation with the greatest timestamp. Specific nodes may remember older generations,
+    // but eventually they forget when their clocks move past the latest generation's timestamp.
+    // The cluster as a whole is only interested in the last generation so restarting nodes may learn what it is.
+    // We assume that generation changes don't happen ``too often'' so every node can learn about a generation
+    // before it is superseded by a newer one which causes nodes to start gossiping the about the newer one.
+    // The assumption follows from the requirement of bootstrapping nodes sequentially.
+    if (!_gen_id || *_gen_id < ts) {
```

Comment:
Is sequential bootstrap really enough to guarantee something like that? Even if nodes are started sequentially, I assume negotiating/disseminating cdc generation info etc takes at least some time. Could not a new node get going during this time, creating an overlap? Is this an issue?
What about nodes crashing & coming back up / decommission-recommission? I assume the distributed mechanism block somehow to finish the current iteration before getting caught up in new things?

Reply:
The system does not prevent the admin from (all) mistakes such as bootstrapping nodes non-sequentially - this is why we want to switch to "Raft-based topology changes", where all decisions about topology changes will be serialized. So yes, it may be possible that the admin does a mistake and tries to join two nodes at the same time, causing two generations to be broadcasted at the same time, causing the one with the smaller timestamp to be learned by only a subset of the nodes (and that would be bad).

There is some small protection: when a node joins it first performs a gossip round and checks the STATUS of other nodes. If it has status=BOOT and sees another node with status=BOOT it will refuse to join (this check happens before we introduce a new generation). But of course races are possible there where the protection fails, it's not perfect. For now we mostly hope that admins follow documentation (which explicitly says to wait between bootstrapping nodes). If they don't, disappearing CDC generations are not their only problem (possible data loss is another).

Reply:
If nodes are bootstrapped sequentially then we're good because the CDC generation broadcast should finish before even streaming starts (we wait for ring_delay after introducing a new generation and before starting streaming data). The next node joins after streaming finishes for this node.

Nodes crashing & restarting may be a problem but it would have to be a coordinated crash. Something like:
- node X creates and starts gossiping a new generation and then immediately crashes
- a minute later it restarts, continues gossiping it, but it means that other nodes in the cluster learn about the generation too late.

This scenario can't happen on bootstrap because if node X was bootstrapping when it crashed then on restart it will create a completely new generation, so it's as if we never created the previous one.
But it can potentially happen on checkAndRepairCdcStreams. It is highly improbable, the node would have to crash precisely after it inserts the generation ID into application state but before it reaches some other node. Still, perhaps it would be good to somehow prevent this, e.g. persisting the timestamp only after it reaches at least one other node (so if we crash in this scenario, we forget about the generation). But this can also break.
There is a solution that preserves consistency under any crash scenarios and is somewhat available but it requires consensus, hence we wait for Raft before implementing it.